Adding support for IIS profiling events in KUDU's process explorer

puneet-gupta opened the pull request:
This code will create a checkbox called "Collect IIS events" next to the START PROFILING button in process explorer in KUDU and this will allow to collect IIS ETW events, along with thread time events for Perfview and they will help in troubleshooting slow performance issues in Azure App Service

dnfclas commented:
@__puneet-gupta__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

dnfclas commented:
__@puneet-gupta__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

safihamid commented:
Please combine all the commits into a single commit for the PRs.

safihamid commented on Kudu.Services.Web/Content/Scripts/ProcessExplorer.js:
```diff
@@ -370,17 +393,20 @@ var Process = (function () {
         }, false);
         $(profilingButton).css("width", "138px");
 
-        if (this._json.is_profile_running) { // highlight button if the profiler has started
```
why is this removed?

davidebbo commented:
@safihamid having multiple commits should not affect your ability to review, since it shows you the overall change to comment on. Only when it comes time to merge it in do we need to worry about squashing.

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
     {
         private const int ProcessExitTimeoutInSeconds = 180;
 
+        private const string SERVICE_PROFILER_AGENT_GUID = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
```
Please rename all const names to Pascal casing like the rest of Kudu files. like ServiceProfilerAgentGuid

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
-        internal static async Task<ProfileResultInfo> StopProfileAsync(int processId, ITracer tracer = null)
+        private static async Task<ProfileResultInfo> StartIISSessionAsync(int processId, int profilingSessionId, ITracer tracer = null)
+        {
+            string arguments = System.Environment.ExpandEnvironmentVariables(string.Format("start {0} /scratchLocation:\"%LOCAL_EXPANDED%\\Temp\" /loadAgent:{1};ServiceProfilerAgent.dll /monitor", profilingSessionId, SERVICE_PROFILER_AGENT_GUID));
```
Are you sure you will need the /monitor switch? why is it needed? we didn't need it for CPU profiling. please look at the help for VSDiagnostics.exe

pharring replied:
You can remove /monitor (it's useful when running from the command line for testing, but not in production)

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
+            // implementing a retry logic because it can so happen
+            // that the intial profiler command may take some time to start
+            // so max we will wait till 9 seconds before giving up
+            int retryCount = 0;
```
There should really be no retry logic here. Tasks can Wait on completion of a task or ContinueWith when a task finishes.

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
+
+            while (retryCount < 2 && profileProcessResponse.StatusCode != HttpStatusCode.OK)
+            {
+                await Task.Delay(3000);
```
This Delay is always happening regardless which is redundant.

puneet-gupta replied:
Just saw your first message and spent hours figuring out how to squash multiple commits in to 1. got that done. Will work on the rest of stuff by tomorrow

puneet-gupta replied:
Just saw your first message and spent hours figuring out how to squash multiple commits in to 1. got that done. Will work on the rest of stuff by tomorrow

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -39,34 +47,96 @@ internal static class ProfileManager
                 // This may fail if we got 2 requests at the same time to start a profiling session
                 // in that case, only 1 will be added and the other one will be stopped.
-                if (!_profilingList.TryAdd(processId, new ProfileInfo(profilingSessionId)))
+                if (!_profilingList.TryAdd(processId, new ProfileInfo(profilingSessionId,iisProfiling)))
```
nit: space after comma.

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
         // The profiling session timeout, this is temp fix before VS2015 Update 1.
         private static readonly TimeSpan _profilingTimeout = TimeSpan.FromMinutes(15);
+
+        private static readonly TimeSpan _profilingIISTimeout = TimeSpan.FromMinutes(3);
```
Can you deploy a standard asp.net site or web api, on a small dedicated worker, enable IIS profiling and load test it with WCAT or some other http stress testing tool for 3 minutes and share the file? Also plz log the timestamp, I want to see how big the file would be and how the VM behaved during that time.
If the file size is too big that even PerfView has problem opening it or if the VM is having Disk Space issues on D drive or the site's temp folder then we must lower the timeout value

puneet-gupta commented:
Addressed following changes in latest commit

1. Changed contant names to pascal naming convention
2. Fixed space before iistracing for the comment 'nit: space after comma'
3. Removed monitor switch from profiler
4. Removed all unwanted logic for retry etc. which I had put earlier for the monitor switch which wasnt necessary.
5. Put back the line in processexplorer.js which got removed accidently

pharring commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
     {
         private const int ProcessExitTimeoutInSeconds = 180;
 
+        private const string ServiceProfilerAgentGuid = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
```
Let's call this ```UserModeCustomProviderAgentGuid```. (While it is bundled in the Service Profiler DLL, that's just for convenience.)
Once you've done that rename, then it frees up the name ```ServiceProfilerAgentGuid``` for what is currently ```ServiceProfilerProviderGuid``` (1C92B...)

pharring commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
 
+        private const string ServiceProfilerAgentGuid = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
+        private const string ServiceProfilerProviderGuid = "1C92BA2A-A990-480F-A02F-40068871CAAC";
+        private const string IISWWWServerGuid = "3A2A4E84-4C21-4981-AE10-3FDA0D9B0F83";
```
I would call this something with the word "Provider" in it. e.g. ```IISWWWServerProviderGuid``` (it's a GUID for an ETW provider)

pharring commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
         private const int ProcessExitTimeoutInSeconds = 180;
 
+        private const string ServiceProfilerAgentGuid = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
+        private const string ServiceProfilerProviderGuid = "1C92BA2A-A990-480F-A02F-40068871CAAC";
```
Let's call this ```DetailedTracingAgentGuid```

pharring commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -39,34 +47,69 @@ internal static class ProfileManager
         }
 
-        internal static async Task<ProfileResultInfo> StopProfileAsync(int processId, ITracer tracer = null)
+        private static async Task<ProfileResultInfo> StartIISSessionAsync(int processId, int profilingSessionId, ITracer tracer = null)
```
How about some comments explaining what's going on here?
e.g. Start a session with the "custom etw provider agent", add in the "detailed tracing agent", turn on the IISWWWServer provider.

safihamid commented on Kudu.Services.Web/Content/Scripts/ProcessExplorer.js:
```diff
@@ -374,13 +397,19 @@ var Process = (function () {
             $(profilingButton).off("click");
             $(profilingButton).attr("title", "Profiling is not supported for Free/Shared sites.");
             $(profilingButton).tooltip().show();
+
```
nit: extra line

safihamid commented on Kudu.Services.Web/Content/Scripts/ProcessExplorer.js:
```diff
@@ -920,26 +949,36 @@ function handleProfilingEvents(e, processId) {
 
+    var iisProfiling = false;
+
+    if (document.getElementById(processId + "-iisProfilingCheck") !=null) {
```
nit: need space, != null

safihamid commented on Kudu.Services.Web/Content/Scripts/ProcessExplorer.js:
```diff
@@ -985,6 +1024,7 @@ function stopProfiling(e, processId) {
 
     Utilities.downloadURL(uri, true);
     e.target.textContent = "Start Profiling";
+
```
nit: extra line.

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
         // The profiling session timeout, this is temp fix before VS2015 Update 1.
         private static readonly TimeSpan _profilingTimeout = TimeSpan.FromMinutes(15);
+
+        private static readonly TimeSpan _profilingIISTimeout = TimeSpan.FromMinutes(3);
```
nit: _profilingIisTimeout

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,24 @@ internal static class ProfileManager
 
+        private const string UserModeCustomProviderAgentGuid = "F5091AA9-80DC-49FF-A7CF-BD1103FE149D";
+        private const string DetailedTracingAgentGuid = "1C92BA2A-A990-480F-A02F-40068871CAAC";
+        private const string IISWWWServerProviderGuid = "3A2A4E84-4C21-4981-AE10-3FDA0D9B0F83";
```
nit: use standard pascal casing for acronyms longer than 2 characters. this better be called IisWebServerProviderGuid

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -39,34 +47,71 @@ internal static class ProfileManager
+                    var profileResponseStatus = await StartIISSessionAsync(processId, profilingSessionId, tracer);
 
-                if (profileProcessResponse.StatusCode != HttpStatusCode.OK)
+                    if (profileResponseStatus.StatusCode != HttpStatusCode.OK)
```
This logic is repeated in both if and else part, move it outside so it is not repeated.

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -39,34 +47,71 @@ internal static class ProfileManager
         }
 
-        internal static async Task<ProfileResultInfo> StopProfileAsync(int processId, ITracer tracer = null)
+        private static async Task<ProfileResultInfo> StartIISSessionAsync(int processId, int profilingSessionId, ITracer tracer = null)
```
rename the method to StartIisSessionAsync

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -83,30 +128,58 @@ internal static class ProfileManager
         {
             return _profilingList.ContainsKey(processId);
         }
+        internal static bool IsIISProfileRunning(int processId)
```
nit: rename to IsIisProfilingRunning

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -83,30 +128,58 @@ internal static class ProfileManager
+
+                if (profilingId != null)
+                    iisProfiling = profilingId.IsIISProfiling;
+
```
nit: extra line

safihamid replied:
nit: the one liner needs still be in curly brackets. 

if (profilingId != null)
{
          iisProfiling = profilingId.IsIISProfiling;
}

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -83,30 +128,58 @@ internal static class ProfileManager
+
+            return iisProfiling;
+        }
+
```
nit: please remove all these extra lines

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -295,15 +386,19 @@ public ProfileResultInfo(HttpStatusCode statusCode, string message)
+           
             public DateTime StartTime { get; set; }
+
+            public bool IsIISProfiling { get; set; }
```
nit: IsIisProfiling

ahmelsayed commented:
What would be the negatives of always including the IIS providers in the profile?

safihamid commented:
@ahmelsayed IIS profiles use /threadTime which log all context switches, block times, ... creating such profiles are very CPU intensive and generate large files. we definitely don;' want to enable them when running CPU profiling which just gets a CPU call stack every millisecond and can be opened in VS for further processing. IIS profiles are for deeper investigations and can be opened in perfView.

safihamid commented:
Can you deploy a standard asp.net site or web api, on a small dedicated worker, enable IIS profiling and load test it with WCAT or some other http stress testing tool for 3 minutes and share the file? Also plz log the timestamp, I want to see how big the file would be and how the VM behaved during that time.
If the file size is too big that even PerfView has problem opening it or if the VM is having Disk Space issues on D drive or the site's temp folder then we must lower the timeout value

safihamid commented on Kudu.Services.Web/Content/Scripts/ProcessExplorer.js:
```diff
@@ -374,13 +397,18 @@ var Process = (function () {
 
+        var iisProfilingCheckbox = Utilities.getCheckbox(this._json.id + "-iisProfilingCheck", "Collect IIS Events")
+       
         if (Process.prototype.WebSiteSku === "Free" || Process.prototype.WebSiteSku === "Shared") {
```
The profiling checkbox should also be disabled(not clickable) for free/shared. Similar to what we are doing for the button.

puneet-gupta replied:
Will do

safihamid commented on Kudu.Services.Web/Content/Scripts/ProcessExplorer.js:
```diff
@@ -920,26 +948,36 @@ function handleProfilingEvents(e, processId) {
         return;
     }
 
+    var iisProfiling = false;
```
As we discussed before, we need to show a message to the user when they check the iis profiling checkbox. The message should have 'ok' and 'cancel' buttons and proceed if 'ok' is clicked.
The message should specify what IIS profiling does, and the fact that 'This is a relatively expensive option' and the file needs to be opened in 'PerfView' with a link to where to download it from. Should also most likely specift the 3 minute timeout or whatever timeout we pick.

puneet-gupta replied:
Will do

puneet-gupta replied:
This is done now !!!

puneet-gupta commented:
Working on the stress test and repro part....should have it ready in a day or two

puneet-gupta commented:
@safihamid - Instead of hardcoding the _profileIisTimeout to 3 minutes (or 1 minute if we end up finalizing it), are we ok to expose an APP_SETTING something like IIS_PROFILING_TIMEOUT_KUDU which an end customer can site and that can override the default timeout ??? 

In some situations, I am thinking that there may be a valid need to increase this timeout and we should have some option for the customer. 

We can always have the MAX to be 15 minutes so that if a customer chooses a higher value than this, it doesn't get honored.

safihamid commented:
@puneet-gupta  That's actually a good idea. we can have the default and then have an option to override if really needed. Also I don't remember if I have ever tested it but what happens to the UI if profiling is started and then they forget to stop it and it times out?

puneet-gupta commented:
@safihamid "Also I don't remember if I have ever tested it but what happens to the UI if profiling is started and then they forget to stop it and it times out?"    

I tested this and found that the UI doesn't auto-update itself. However if someone clicks REFRESH or Refreshes the browser after the timeout then the button status is correctly set to Start Profiling

Shall I also implement the logic for IIS_PROFILING_TIMEOUT_KUDU  and does this name look OK?

safihamid commented:
@puneet-gupta how about  IIS_PROFILING_TIMEOUT_OVERRIDE_IN_SECONDS?

puneet-gupta commented:
seems to long, how about **IIS_PROFILING_TIMEOUT_IN_SECONDS**

safihamid commented:
good! :-)

puneet-gupta commented:
cool , will update this tomorrow

puneet-gupta commented:
@safihamid also shared a sample ETW trace with you over email to have a quick review

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -35,6 +35,16 @@ internal static class ProfileManager
         private static string _processName = System.Environment.ExpandEnvironmentVariables("%SystemDrive%\\msvsmon\\profiler\\VSStandardCollector.Dev14.exe");
 
+        static ProfileManager()
+        {
```
Please combine this logic with private static readonly TimeSpan _profilingIisTimeout = TimeSpan.FromMinutes(3);
and remove this static assignment.

puneet-gupta replied:
Sorry not sure if I am understanding properly, you want me to remove this line 

 `private static readonly TimeSpan _profilingIisTimeout = TimeSpan.FromMinutes(3);`

or the static constructor ?
```
static ProfileManager()
        {
            int iisProfilingTimeout = GetIisProfilingTimeout();

            if (iisProfilingTimeout > 0)
            {
                _profilingIisTimeout = TimeSpan.FromSeconds(iisProfilingTimeout);
            }
        }
```

safihamid replied:
@puneet-gupta plz remove the assignment here: private static readonly TimeSpan _profilingIisTimeout = TimeSpan.FromMinutes(3);
and add that logic to the static constructor.

safihamid commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -104,6 +116,26 @@ internal static class ProfileManager
+            {
+                if (Int32.TryParse(iisProfilingTimeoutInSeconds, out timeout))
+                {
+                    if (timeout>(60*15)) // making sure that no configures it more than 15 minutes
```
Instead of 60*15 please use 60*_profilingTimeout, also fix the spacing and comment

puneet-gupta replied:
not following...This is not actually a multiple of _profliingTimeout. This is an upper limit of _profilingTimeout....Looks like your comment is missing some characters...

safihamid replied:
@puneet-gupta _profilingTimeout is already set to 15 minutes. So I am suggesting to use that value instead. so replcae 60 x 15 with 60 x _profilingTimeout

safihamid commented on Kudu.Contracts/Diagnostics/ProcessInfo.cs:
```diff
@@ -30,6 +30,12 @@ public class ProcessInfo : INamedObject
+        [JsonProperty(PropertyName = "is_iis_profile_running", DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public bool IsIisProfileRunning { get; set; }
+
+        [JsonProperty(PropertyName = "iis_profile_timeout_seconds", DefaultValueHandling = DefaultValueHandling.Ignore)]
```
I would say let's call this: iis_profile_timeout_in_seconds to be compat with rest of the code

safihamid commented on Kudu.Services.Web/Content/Scripts/ProcessExplorer.js:
```diff
@@ -943,23 +960,73 @@ function searchForHandle() {
+
+            $(divConfirm).dialog({
+                resizable: false,
+                height: 330,
```
Nit: not a javascript expert but are we supposed to hard-code the sizes here? how does it show on different resolutions of hard coded? maybe check rest of Kudu, if they are same then ok.

puneet-gupta commented:
@safihamid  Regarding the specifying height and width, I do see hard-coded height and width at some places in code. If I am leaving this to AUTO then the dialog is looking a bit ugly as there is quite some information displayed in the message box.

It seems I have addressed the issues but please let me know if I missed something (It seems some comments are missing for some reason but I remember them reading here)

pharring commented on Kudu.Services/Diagnostics/ProfileManager.cs:
```diff
@@ -18,16 +18,29 @@ internal static class ProfileManager
         // The profiling session timeout, this is temp fix before VS2015 Update 1.
         private static readonly TimeSpan _profilingTimeout = TimeSpan.FromMinutes(15);
+
+        private static readonly TimeSpan _profilingIisTimeout = TimeSpan.MinValue;
```
You can write
```C#
private static readonly TimeSpan _profilingIisTimeout = GetIisProfilingTimeout();
```

and then eliminate the static constructor entirely.

puneet-gupta replied:
Thanks Paul, checked in this change !!!

davidebbo commented:
I merged the change. Thanks @puneet-gupta for your contrib!

puneet-gupta commented:
Thank you Hamid , Paul and David for letting me contribute to this project !!!! It was a great experience for me and I am sure it will help support and customers tremendously....

Do we know when this available to all in Kudu ?

davidebbo commented:
Hopefully we'll deployit next week.
